pico_w/wifi/access_point/dhcpserver/dhcpserver.c

```diff
@@ -288,7 +288,7 @@ static void dhcp_server_process(void *arg, struct udp_pcb *upcb, struct pbuf *p,
     opt_write_u32(&opt, DHCP_OPT_IP_LEASE_TIME, DEFAULT_LEASE_TIME_S);
     *opt++ = DHCP_OPT_END;
     struct netif *nif = ip_current_input_netif();
-    dhcp_socket_sendto(&d->udp, nif, &dhcp_msg, opt - (uint8_t *)&dhcp_msg, 0xffffffff, PORT_DHCP_CLIENT);
+    dhcp_socket_sendto(&upcb, nif, &dhcp_msg, opt - (uint8_t *)&dhcp_msg, 0xffffffff, PORT_DHCP_CLIENT);
```

Comment:
I don't have anything handy to debug it more thoroughly , pico just hangs the moment it tries to send back udp packet.

What do you mean by `And I see d->udp is always equal to updb`? The first pcb is used to set up callbacks (binding, etc.) then the recv callback receives another pcb, which is never in use, because packet is being sent back via the first pcb instance.

Signature for recv callback: `(void *arg, struct udp_pcb *pcb, struct pbuf *p, const ip_addr_t *addr, u16_t port)`.

Reply:
I put the following before the call and the assert never failed.
assert(d && d->udp == upcb);

Reply:
So do I and it stops execution.

![image](https://github.com/raspberrypi/pico-examples/assets/33105795/3644c1b1-4c1c-4f8d-acfa-dd47f1a611af)